# UPLOADING THE DATASET

**I have implemented Logistic Regression on the dataset. This is my first ever written kernel and I hope that it will be valuable to others. The tasks I have done will be described as the code goes on.**

Libaries are imported for the Code


In [6]:
import numpy as np
import pandas as pd

df = pd.read_csv("../input/logistic-regression-heart-disease-prediction/framingham_heart_disease.csv")
df.head()

male  age  education  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0     1   39        4.0              0         0.0     0.0                0   
1     0   46        2.0              0         0.0     0.0                0   
2     1   48        1.0              1        20.0     0.0                0   
3     0   61        3.0              1        30.0     0.0                0   
4     0   46        3.0              1        23.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             0         0    195.0  106.0   70.0  26.97       80.0     77.0   
1             0         0    250.0  121.0   81.0  28.73       95.0     76.0   
2             0         0    245.0  127.5   80.0  25.34       75.0     70.0   
3             1         0    225.0  150.0   95.0  28.58       65.0    103.0   
4             0         0    285.0  130.0   84.0  23.10       85.0     85.0   

   TenYearCHD  
0           0  
1           0  
2           0  
3           1  
4           0

**The total missing values from the dataset are calculated and shown on the screen.**

In [7]:
df.isnull().sum()

male                 0
age                  0
education          105
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

**This shows the CigsPerDay column and displays all the Null values in the column.**

# CLEANING OF THE DATASET

In [8]:
series = pd.isnull(df['cigsPerDay'])
df[series]

male  age  education  currentSmoker  cigsPerDay  BPMeds  \
131      1   43        4.0              1         NaN     0.0   
139      1   49        4.0              1         NaN     0.0   
1046     0   49        1.0              1         NaN     0.0   
1292     1   42        3.0              1         NaN     0.0   
1347     0   58        4.0              1         NaN     0.0   
1451     1   54        1.0              1         NaN     0.0   
1497     1   55        1.0              1         NaN     0.0   
1610     0   61        1.0              1         NaN     0.0   
1625     0   49        2.0              1         NaN     0.0   
1870     0   47        2.0              1         NaN     0.0   
1963     1   45        3.0              1         NaN     0.0   
1980     0   60        2.0              1         NaN     0.0   
2405     0   49        1.0              1         NaN     0.0   
2513     1   42        1.0              1         NaN     0.0   
2542     1   57        NaN              1         NaN     0.0   
3021     1   53        2.0              1         NaN     0.0   
3034     0   57        NaN              1         NaN     0.0   
3094     0   49        1.0              1         NaN     0.0   
3106     0   47        3.0              1         NaN     0.0   
3108     1   39        2.0              1         NaN     0.0   
3156     1   37        3.0              1         NaN     0.0   
3177     1   45        2.0              1         NaN     0.0   
3309     1   58        2.0              1         NaN     0.0   
3432     0   55        1.0              1         NaN     0.0   
3579     1   42        1.0              1         NaN     0.0   
3715     0   57        2.0              1         NaN     0.0   
3847     0   41        1.0              1         NaN     0.0   
3924     0   52        2.0              1         NaN     0.0   
3942     0   39        3.0              1         NaN     0.0   

      prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
131                 0             0         0    222.0  109.5   69.0  25.50   
139                 0             0         0    256.0  127.5   81.5  28.21   
1046                0             0         0    280.0  120.0   80.0  22.33   
1292                0             0         0    225.0  122.5   80.0  25.54   
1347                0             1         0    270.0  195.0  117.5  23.35   
1451                0             0         0    219.0  110.0   72.0  26.05   
1497                0             0         0    214.0  132.5   85.5  29.25   
1610                0             1         0    356.0  168.0   98.0  27.30   
1625                0             1         0    233.0  158.0  102.0  25.31   
1870                0             0         0    365.0  127.0   76.0  24.44   
1963                0             1         0    170.0  145.5   99.0  26.74   
1980                0             0         0    228.0  112.0   74.0  24.51   
2405                0             0         0    252.0  123.0   69.0  21.45   
2513                0             0         0    226.0  119.0   80.0  25.29   
2542                0             0         0    223.0  107.5   72.5  24.74   
3021                0             0         0    276.0  130.0   86.0  24.21   
3034                0             0         0    229.0  115.0   69.0  24.43   
3094                0             1         0    214.0  172.0  111.0  40.51   
3106                0             0         0    321.0  132.0   88.0  28.14   
3108                0             0         0    285.0  121.0   82.0  27.62   
3156                0             0         0    188.0  123.5   77.0  26.62   
3177                0             0         0    248.0  121.0   72.0  27.88   
3309                0             0         0    235.0  127.5   76.0  21.02   
3432                0             1         0    213.0  163.0   91.0  28.66   
3579                0             0         0    196.0  123.0   73.0  22.06   
3

The advantage of above table shown is that the 'currentSmoker' column has value of 1 for all 'cigsPerDay' null values. This helps in conveying that all missing values of cigsPerDay would not be zero.

* **As Education does not play a factor in Heart Attack, I have dropped that column**
* **Moreover, I dropped currentSmoker column too as cigsPerDay column already denotes that the person is a Smoker, that's just extra information which is not useful**

In [9]:
data = df.drop(['currentSmoker','education'], axis = 'columns')
data.head()

male  age  cigsPerDay  BPMeds  prevalentStroke  prevalentHyp  diabetes  \
0     1   39         0.0     0.0                0             0         0   
1     0   46         0.0     0.0                0             0         0   
2     1   48        20.0     0.0                0             0         0   
3     0   61        30.0     0.0                0             1         0   
4     0   46        23.0     0.0                0             0         0   

   totChol  sysBP  diaBP    BMI  heartRate  glucose  TenYearCHD  
0    195.0  106.0   70.0  26.97       80.0     77.0           0  
1    250.0  121.0   81.0  28.73       95.0     76.0           0  
2    245.0  127.5   80.0  25.34       75.0     70.0           0  
3    225.0  150.0   95.0  28.58       65.0    103.0           1  
4    285.0  130.0   84.0  23.10       85.0     85.0           0

In [10]:
cigs = data['cigsPerDay']
cigs.head()

0     0.0
1     0.0
2    20.0
3    30.0
4    23.0
Name: cigsPerDay, dtype: float64

**I have calculated the mean of people smoking a cigaratte per day**

In [11]:
cig = cigs.mean()

I rounded that value to the neartest integer

In [12]:
import math
integer_value = math.floor(cig)
integer_value

9

**I filled the null values in 'cigsPerDay' with the mean of the Cigarattes smoked by a person in a day.**

In [15]:
cigs.fillna(integer_value, inplace = True)

In [16]:
data.isnull().sum()

male                 0
age                  0
cigsPerDay           0
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

**I dropped the rest of the rows of the null values**

In [18]:
data.dropna( axis = 0, inplace = True)

**This shows that there are no Null value anymore in the DataFrame**

In [20]:
data.isnull().sum()

male               0
age                0
cigsPerDay         0
BPMeds             0
prevalentStroke    0
prevalentHyp       0
diabetes           0
totChol            0
sysBP              0
diaBP              0
BMI                0
heartRate          0
glucose            0
TenYearCHD         0
dtype: int64

# ANALYZING THE DATASET

In [21]:
data.shape

(3774, 14)

**Created a separate DataFrame for the people having chances of heart attack**

In [24]:
Heart_Attack = data[data.TenYearCHD == 1]
Heart_Attack.head()

male  age  cigsPerDay  BPMeds  prevalentStroke  prevalentHyp  diabetes  \
3      0   61        30.0     0.0                0             1         0   
6      0   63         0.0     0.0                0             0         0   
15     0   38        20.0     0.0                0             1         0   
17     0   46        20.0     0.0                0             0         0   
25     1   47        20.0     0.0                0             0         0   

    totChol  sysBP  diaBP    BMI  heartRate  glucose  TenYearCHD  
3     225.0  150.0   95.0  28.58       65.0    103.0           1  
6     205.0  138.0   71.0  33.11       60.0     85.0           1  
15    221.0  140.0   90.0  21.35       95.0     70.0           1  
17    291.0  112.0   78.0  23.38       80.0     89.0           1  
25    294.0  102.0   68.0  24.18       62.0     66.0           1

**Created another separate DataFrame for people having low chances of heart attack**

In [25]:
No_Heart_Attack = data[data.TenYearCHD == 0]
No_Heart_Attack.head()

male  age  cigsPerDay  BPMeds  prevalentStroke  prevalentHyp  diabetes  \
0     1   39         0.0     0.0                0             0         0   
1     0   46         0.0     0.0                0             0         0   
2     1   48        20.0     0.0                0             0         0   
4     0   46        23.0     0.0                0             0         0   
5     0   43         0.0     0.0                0             1         0   

   totChol  sysBP  diaBP    BMI  heartRate  glucose  TenYearCHD  
0    195.0  106.0   70.0  26.97       80.0     77.0           0  
1    250.0  121.0   81.0  28.73       95.0     76.0           0  
2    245.0  127.5   80.0  25.34       75.0     70.0           0  
4    285.0  130.0   84.0  23.10       85.0     85.0           0  
5    228.0  180.0  110.0  30.30       77.0     99.0           0

**This groups the data on the basis of 'TenYearCHD' indicating the overall dependence of columns on 'TenYearCHD'**

In [26]:
data.groupby('TenYearCHD').mean()

male        age  cigsPerDay    BPMeds  prevalentStroke  \
TenYearCHD                                                               
0           0.425312  48.740313    8.723750  0.024063         0.004063   
1           0.557491  54.233449   10.574913  0.064460         0.013937   

            prevalentHyp  diabetes     totChol       sysBP      diaBP  \
TenYearCHD                                                              
0               0.276250  0.020625  235.347187  130.308437  82.182969   
1               0.506969  0.062718  246.311847  143.827526  87.161150   

                  BMI  heartRate    glucose  
TenYearCHD                                   
0           25.662769  75.609688  80.665000  
1           26.644094  76.378049  88.827526

**As 'diaBP', 'BMI', 'heartRate' have values in the similar zone and almost equal to one another for 1 or 0 values for 'TenYearCHD', they are dropped.**

In [27]:
final = data.drop(['diaBP','BMI','heartRate'], axis = 'columns')

In [28]:
No_Heart_Attack = final[final.TenYearCHD == 0]
No_Heart_Attack.head()

male  age  cigsPerDay  BPMeds  prevalentStroke  prevalentHyp  diabetes  \
0     1   39         0.0     0.0                0             0         0   
1     0   46         0.0     0.0                0             0         0   
2     1   48        20.0     0.0                0             0         0   
4     0   46        23.0     0.0                0             0         0   
5     0   43         0.0     0.0                0             1         0   

   totChol  sysBP  glucose  TenYearCHD  
0    195.0  106.0     77.0           0  
1    250.0  121.0     76.0           0  
2    245.0  127.5     70.0           0  
4    285.0  130.0     85.0           0  
5    228.0  180.0     99.0           0

In [29]:
Heart_Attack = final[final.TenYearCHD == 1]
Heart_Attack.head()

male  age  cigsPerDay  BPMeds  prevalentStroke  prevalentHyp  diabetes  \
3      0   61        30.0     0.0                0             1         0   
6      0   63         0.0     0.0                0             0         0   
15     0   38        20.0     0.0                0             1         0   
17     0   46        20.0     0.0                0             0         0   
25     1   47        20.0     0.0                0             0         0   

    totChol  sysBP  glucose  TenYearCHD  
3     225.0  150.0    103.0           1  
6     205.0  138.0     85.0           1  
15    221.0  140.0     70.0           1  
17    291.0  112.0     89.0           1  
25    294.0  102.0     66.0           1

In [30]:
final.groupby('TenYearCHD').mean()

male        age  cigsPerDay    BPMeds  prevalentStroke  \
TenYearCHD                                                               
0           0.425312  48.740313    8.723750  0.024063         0.004063   
1           0.557491  54.233449   10.574913  0.064460         0.013937   

            prevalentHyp  diabetes     totChol       sysBP    glucose  
TenYearCHD                                                             
0               0.276250  0.020625  235.347187  130.308437  80.665000  
1               0.506969  0.062718  246.311847  143.827526  88.827526

# PREPARING THE MODEL

In [32]:
X = final[['male','age','cigsPerDay','BPMeds','prevalentStroke','prevalentHyp','diabetes','totChol','sysBP','glucose']]

In [33]:
y = final['TenYearCHD']

In [34]:
X

male  age  cigsPerDay  BPMeds  prevalentStroke  prevalentHyp  diabetes  \
0        1   39         0.0     0.0                0             0         0   
1        0   46         0.0     0.0                0             0         0   
2        1   48        20.0     0.0                0             0         0   
3        0   61        30.0     0.0                0             1         0   
4        0   46        23.0     0.0                0             0         0   
...    ...  ...         ...     ...              ...           ...       ...   
4231     1   58         0.0     0.0                0             1         0   
4232     1   68         0.0     0.0                0             1         0   
4233     1   50         1.0     0.0                0             1         0   
4234     1   51        43.0     0.0                0             0         0   
4237     0   52         0.0     0.0                0             0         0   

      totChol  sysBP  glucose  
0       195.0  106.0     77.0  
1       250.0  121.0     76.0  
2       245.0  127.5     70.0  
3       225.0  150.0    103.0  
4       285.0  130.0     85.0  
...       ...    ...      ...  
4231    187.0  141.0     81.0  
4232    176.0  168.0     79.0  
4233    313.0  179.0     86.0  
4234    207.0  126.5     68.0  
4237    269.0  133.5    107.0  

[3774 rows x 10 columns]

In [35]:
y

0       0
1       0
2       0
3       1
4       0
       ..
4231    0
4232    1
4233    1
4234    0
4237    0
Name: TenYearCHD, Length: 3774, dtype: int64

**I have split the model in 20-80 test and train size with a given random state so that the output doesnt differ everytime I execute the program**

In [37]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.20, random_state = 99)

**Logistic Regression is implemented**

In [39]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()

In [40]:
model.fit(X_train, y_train)

/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [41]:
model.score(X_test,y_test)

0.8556291390728477

**The model is 85.5% accurate. If I change the random state, it still gives the output in the range of 83-86%.**